### ISSUE TITLE
About the way translated language names are handled

### ISSUE THREAD

**Issue by BogdanAlexandru** [HospitalRun/hospitalrun-frontend#1245]
While working on the Romanian translation I stumbled upon this:

```js
// the French localization
languages: {
    en: 'Anglais',
    fr: 'Français',
    es: 'Espagnol',
    de: 'Allemand',
    // ...
  }
```

Each localization specifies its own language-specific names for the other languages. If one entry is missing, it shows up in the app as an alternative, "missing language" string.

But my question is rather about whether or not this approach helps the user. In all of the programs I've used, language names are not current-language specific. So for the French translation above, the fields would look like:

```js
// the French localization
languages: {
    en: 'English'
    fr: 'Français',
    es: 'Español',
    de: 'Deutsche',
    ro: 'Română',
    // and so on
  }
```

This is useful because if the program runs in a language that the user does not understand (or he accidentally picks an incorrect language), he can easily find his own language in the list (since it uses a familiar name). For some languages, the difference is quite large e.g. for French: "Allemand" vs "Deutsche", so I can see there being some difficulty. For other languages, the difference can be quite VERY LARGE. For example, when trying to switch to Romanian, I accidentally chose Urdu. I have no idea which one of these is Romanian, or even English, so I could go back:

![Imgur](https://i.imgur.com/FzdYkll.png)

While at it, would something like this kind of language selector not be more intuitive?
![img](https://alistapart.com/d/Column-Antoine/F1_blabla.png)
I can much quicker spot the flag of my country/other familiar country (language) in such a list, compared to a plain text list. Not to mention, it also looks like great eyecandy for very little effort.

Taking it out of the settings dropdown and displaying the flag directly would also remove the requirement for the current user to understand the currently selected language in order to tell where the language dropdown lies. This is not *that* big of an issue, especially given the really simple dropdown UI that HR+ currently has, but it's still some text sitting between the user and his preferred language, that he may not be able to understand.

![Imgur](https://i.imgur.com/VD4GJNU.png)

Flags shown directly are pretty universal and would ameliorate this issue.

Thoughts?

**Comment by BogdanAlexandru** [comment]
Related, from a programmer perspective: from what I noticed in previous PRs, language name labels are entered manually (for example, the Turkish translation for "French"). In order to have a full Romanian translation, I need to also update the other locales to contain their translation of "Romanian" in the dropdown. While this is not a difficult task, and can be checked for correctness using Google, it's still a change that would technically require a review from speakers of said languages to see if my translation is correct. And I would definitely appreciate a review in the case of, say, Urdu, which is different from the languages I speak in all regards and not even Google could convince me I'm using the proper translation.

**Comment by tangollama** [comment]
@BogdanAlexandru I think it's a good feature change. Would you like me to turn this into a Help Wanted request (I'll be more specific on what we want the change to be) or would you like to take a crack at it yourself? Let me know.

**Comment by BogdanAlexandru** [comment]
@tangollama This can be split into two issues, I think. For starters, I can change the dropdown to display languages like:

```  
    English
    Français
    Español
    Deutsche
    Română
```
regardless of the currently selected language. I can work on that now.

The 2nd issue would be about adding a flag for each language and moving the selector outside the menu, and that could be a Help Wanted with a more precise design specification.

**Comment by arr99** [comment]
康福中国ID99用户名使用4年了突然掉线在登陆用户名无效

**Comment by arr99** [comment]
康福中国ID99用户名使用4年了突然掉线在登陆用户名无效

**Comment by arr99** [comment]
亲爱的康福领导你好。我是之前99马甲的主任康福99马甲我使用了4年我是大概2013.8月分左右使用的我是在康福中国中国区的 经销商那里购买的，1250美金购买的，我使用了4年都没有什么问题，在2016年7月分的一天，我在康福中国一个小房间和交个朋友在唱歌玩，我的99马甲突然掉线我在登陆用户名无效，我就给你们康福联系，你们也帮忙给我找康福99马甲了。我登陆不了但是邮箱还是我的我用邮箱找回密码就发过来消息说邮箱以被拦截，这个事情我把截图也发给你们了。康福每天叫我输密码什么的。说着20天就过去了突然99马甲登陆了是一个30马甲更改的，我就问他99马甲什么情况他说他是康福买的300美金我要喜欢那500美金，你们康福那时间给我的我99马甲没问题，我就没有在和那30更改99人联系了。说着一个月回去了突然我看见99马甲直接登陆了。是一个中国人在使用，我问那中国人99马甲他说他朋友在康福够买的，我给你们康福说这个事情你们就在也不回来我话了，我心里很生气，在这个时间有一个美国人说他可以帮忙找回来，我说可以，跟着就把马甲的的原因从头 到尾说一遍。三个月过去了我99马甲没有找回来我还花了300美金买软件。说着又一个软件公司说忙吧找马甲，我又花200多美金，到现在马甲没有音信，康福领导我求求你们帮忙给我查看看看我的马甲为什么就这样不叫我使用了。谢谢康福领导 99啊辉

**Comment by arr99** [comment]
亲爱的康福领导你好。我是之前99马甲的主任康福99马甲我使用了4年我是大概2013.8月分左右使用的我是在康福中国中国区的 经销商那里购买的，1250美金购买的，我使用了4年都没有什么问题，在2016年7月分的一天，我在康福中国一个小房间和交个朋友在唱歌玩，我的99马甲突然掉线我在登陆用户名无效，我就给你们康福联系，你们也帮忙给我找康福99马甲了。我登陆不了但是邮箱还是我的我用邮箱找回密码就发过来消息说邮箱以被拦截，这个事情我把截图也发给你们了。康福每天叫我输密码什么的。说着20天就过去了突然99马甲登陆了是一个30马甲更改的，我就问他99马甲什么情况他说他是康福买的300美金我要喜欢那500美金，你们康福那时间给我的我99马甲没问题，我就没有在和那30更改99人联系了。说着一个月回去了突然我看见99马甲直接登陆了。是一个中国人在使用，我问那中国人99马甲他说他朋友在康福够买的，我给你们康福说这个事情你们就在也不回来我话了，我心里很生气，在这个时间有一个美国人说他可以帮忙找回来，我说可以，跟着就把马甲的的原因从头 到尾说一遍。三个月过去了我99马甲没有找回来我还花了300美金买软件。说着又一个软件公司说忙吧找马甲，我又花200多美金，到现在马甲没有音信，康福领导我求求你们帮忙给我查看看看我的马甲为什么就这样不叫我使用了。谢谢康福领导 99啊辉

**Comment by arr99** [comment]
亲爱的康福领导你好。我是之前99马甲的主任康福99马甲我使用了4年我是大概2013.8月分左右使用的我是在康福中国中国区的 经销商那里购买的，1250美金购买的，我使用了4年都没有什么问题，在2016年7月分的一天，我在康福中国一个小房间和交个朋友在唱歌玩，我的99马甲突然掉线我在登陆用户名无效，我就给你们康福联系，你们也帮忙给我找康福99马甲了。我登陆不了但是邮箱还是我的我用邮箱找回密码就发过来消息说邮箱以被拦截，这个事情我把截图也发给你们了。康福每天叫我输密码什么的。说着20天就过去了突然99马甲登陆了是一个30马甲更改的，我就问他99马甲什么情况他说他是康福买的300美金我要喜欢那500美金，你们康福那时间给我的我99马甲没问题，我就没有在和那30更改99人联系了。说着一个月回去了突然我看见99马甲直接登陆了。是一个中国人在使用，我问那中国人99马甲他说他朋友在康福够买的，我给你们康福说这个事情你们就在也不回来我话了，我心里很生气，在这个时间有一个美国人说他可以帮忙找回来，我说可以，跟着就把马甲的的原因从头 到尾说一遍。三个月过去了我99马甲没有找回来我还花了300美金买软件。说着又一个软件公司说忙吧找马甲，我又花200多美金，到现在马甲没有音信，康福领导我求求你们帮忙给我查看看看我的马甲为什么就这样不叫我使用了。谢谢康福领导 99啊辉

**Comment by arr99** [comment]
亲爱的康福领导你好。我是之前99马甲的主任康福99马甲我使用了4年我是大概2013.8月分左右使用的我是在康福中国中国区的 经销商那里购买的，1250美金购买的，我使用了4年都没有什么问题，在2016年7月分的一天，我在康福中国一个小房间和交个朋友在唱歌玩，我的99马甲突然掉线我在登陆用户名无效，我就给你们康福联系，你们也帮忙给我找康福99马甲了。我登陆不了但是邮箱还是我的我用邮箱找回密码就发过来消息说邮箱以被拦截，这个事情我把截图也发给你们了。康福每天叫我输密码什么的。说着20天就过去了突然99马甲登陆了是一个30马甲更改的，我就问他99马甲什么情况他说他是康福买的300美金我要喜欢那500美金，你们康福那时间给我的我99马甲没问题，我就没有在和那30更改99人联系了。说着一个月回去了突然我看见99马甲直接登陆了。是一个中国人在使用，我问那中国人99马甲他说他朋友在康福够买的，我给你们康福说这个事情你们就在也不回来我话了，我心里很生气，在这个时间有一个美国人说他可以帮忙找回来，我说可以，跟着就把马甲的的原因从头 到尾说一遍。三个月过去了我99马甲没有找回来我还花了300美金买软件。说着又一个软件公司说忙吧找马甲，我又花200多美金，到现在马甲没有音信，康福领导我求求你们帮忙给我查看看看我的马甲为什么就这样不叫我使用了。谢谢康福领导 99啊辉

**Comment by arr99** [comment]
亲爱的康福领导你好。我是之前99马甲的主任康福99马甲我使用了4年我是大概2013.8月分左右使用的我是在康福中国中国区的 经销商那里购买的，1250美金购买的，我使用了4年都没有什么问题，在2016年7月分的一天，我在康福中国一个小房间和交个朋友在唱歌玩，我的99马甲突然掉线我在登陆用户名无效，我就给你们康福联系，你们也帮忙给我找康福99马甲了。我登陆不了但是邮箱还是我的我用邮箱找回密码就发过来消息说邮箱以被拦截，这个事情我把截图也发给你们了。康福每天叫我输密码什么的。说着20天就过去了突然99马甲登陆了是一个30马甲更改的，我就问他99马甲什么情况他说他是康福买的300美金我要喜欢那500美金，你们康福那时间给我的我99马甲没问题，我就没有在和那30更改99人联系了。说着一个月回去了突然我看见99马甲直接登陆了。是一个中国人在使用，我问那中国人99马甲他说他朋友在康福够买的，我给你们康福说这个事情你们就在也不回来我话了，我心里很生气，在这个时间有一个美国人说他可以帮忙找回来，我说可以，跟着就把马甲的的原因从头 到尾说一遍。三个月过去了我99马甲没有找回来我还花了300美金买软件。说着又一个软件公司说忙吧找马甲，我又花200多美金，到现在马甲没有音信，康福领导我求求你们帮忙给我查看看看我的马甲为什么就这样不叫我使用了。谢谢康福领导 99啊辉

**Comment by BogdanAlexandru** [comment]
One thing worth considering when deciding whether we want flags or not: flags represent countries, and one country may have multiple languages being spoken. Case in point: Belgium w/ Dutch, French and German. Canada with French and English.

**Comment by MhdTlb** [comment]
Can I work on this?

**Comment by ben-muldrow** [comment]
@tangollama Starting to work on this as my first task in HospitalRun.

**Comment by ben-muldrow** [comment]
@BogdanAlexandru One alternative to flags would be ISO 639-1 language codes. For instance 
```
EN English
FR Français
```
etc. Those are culture-independent and language-specific.

Furthermore, very tiny potential issue but it might not be intuitive where one can go to change the language if stuck in a language they don't know. The image referenced by BogdanAlexandru at the bottom of this post illustrates this nicely. If stuck in a language with characters that are unknown to the user, they might not know to visit this dropdown menu. One solution would be a language icon similar to [this](https://thenounproject.com/search/?q=language&i=1048085).

![image](https://user-images.githubusercontent.com/14099028/35399548-7e64fb1e-01c2-11e8-9832-7699dbdc532e.png)

**Comment by davidackroyd99** [comment]
I can work on the flag/language drop down if that's what's required

**Comment by fox1t** [comment]
@BogdanAlexandru I agree 100% with you. It would be fully implemented like this in 2.0.0 release.
